[Vectorized][UDF] support java-udaf

zhangstar333 opened the pull request:
# Proposed changes

Issue Number: close #8389

## Problem Summary:
This feature is used to support Java udaf. Users can create aggregate functions by self,
For example, we create Java udaf function by code below.

```
CREATE AGGREGATE FUNCTION simple_add (int) RETURNS int PROPERTIES (
    "file"="file:///pathTo/java-udaf.jar",
    "symbol"="org.apache.doris.udf.SimpleAdd",
    "type"="JAVA_UDF"
);
```

**file:** indicate where user file is.
**symbol:** for java udaf means udaf class in this jar.
**type:** indicate this function is a java udf.

Users may have some necessary conditions 
when implementing aggregate functions by self,
for example:

```
class SimpleAdd {
    public static class State { 
        some variables: 
        example count agg: maybe a long type counter;
    }

    public State create() {
        ....... if need do some.......
        return new State();
    }

    public void destroy(State s) {
        ....... if need do some destroy work.......
    }

    public void add(State s, Type val, [Type val2, .....]) {
        do update work here;
        like count agg function: you need plus 1
   }

    public void serialize(State s, DataOutputStream out) {
        put you add result into buffer to serialize
   }

  public void deserialize(State state, DataInputStream in) {
      deserialize get data from buffer before you put
  }

    public void merge(State s, State other) {
       merge result from other state
    }

    public Type getValue(State s) {
          return finally result
   }
}

```

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

zhangstar333 committed:
[Vectorized][UDF] support java-udaf

rebase to master

fix formatter error

refactor add function

change some code according review

add doc file about java-udaf

refactor serialize and deserialize function

BiteTheDDDDt commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,392 @@
+    }
+
+    void add(const IColumn** columns, size_t row_num, const DataTypes& argument_types) const {
+        JNIEnv* env = nullptr;
```
we can wrap a function `JNIEnv* get_env(string when)`

BiteTheDDDDt commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,392 @@
+            LOG(WARNING) << "Java-Udaf get error when merge: " << ret_code.get_error_msg();
+        }
+        serialize_data = rhs.serialize_data;
+        long len = serialize_data.length();
```
Should `long` be used here? `long` has different meaning in `c++` and `java`.

BiteTheDDDDt commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,392 @@
+             Arena*) const override {
+        Status status = data(place).init_udaf(fn_);
+        if (!status.ok()) {
+            LOG(WARNING) << " in add pressore  of init udaf has error: " << status.get_error_msg();
```
multiple consecutive spaces

BiteTheDDDDt commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,392 @@
+                     << "$0 now !";                                                          \
+    }
+#endif
+            EVALUATE_JAVA_UDAF;
```
Do you just want a lambda function?

morningman commented:
Hi @zhangstar333 , could you explain more about what does this PR do?
Because we already supported part of JAVA udf feature, so you need to explain what's new in your PR.

HappenLee commented:
need also modify the doc，please

HappenLee commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,361 @@
+    }
+
+private:
+    TFunction fn_;
```
`_fn`

HappenLee commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,361 @@
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
+             Arena*) const override {
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(fn_));
```
why not do init in `create` function and delete the `first_init`?

zhangstar333 replied:
Have test in create function do init operator, but it's unable to get jvm

HappenLee commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    Status get(IColumn& to, const DataTypePtr& result_type) const {
+        to.insert_default();
```
rethink the `insert_default()` of column string

HappenLee commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    Status merge(const AggregateJavaUdafData& rhs) {
+        if (rhs.first_init) {
```
here is werid of `first_init`, rethink the logic

HappenLee commented on fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java:
```diff
@@ -0,0 +1,539 @@
+    }
+
+    private void allocateInputObjects(long row) throws UdfRuntimeException {
+        inputObjects = new Object[argTypes.length];
```
why not return inputObjects?

HappenLee commented on fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java:
```diff
@@ -0,0 +1,539 @@
+                        inputArgs[i + 1] = null;
+                    }
+                }
+                inputArgs[0] = stateObj;
```
only need set one time?

HappenLee commented on fe/java-udf/src/main/java/org/apache/doris/udf/UdafExecutor.java:
```diff
@@ -0,0 +1,539 @@
+        }
+    }
+
+    private URLClassLoader getClassLoader(String jarPath) throws MalformedURLException {
```
same as method of `UdfExecutor` only keep one, please

Gabriel39 commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,363 @@
+private:
+    static const int32_t INITIAL_RESERVED_BUFFER_SIZE = 1024;
+    // TODO: we need a heuristic strategy to increase buffer size for variable-size output.
+    static inline int32_t IncreaseReservedBufferSize(int n) {
```
Move this function to another common file and call it in UDAF and UDF

zhangstar333 committed:
checkout env after calling method
remove test log
change init_udaf to create function
be formatter
change to trigger be test
remove first_init variable

Gabriel39 commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,358 @@
+    void create(AggregateDataPtr __restrict place) const override {
+        new (place) Data(argument_types.size());
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(_fn));
```
use `RETURN_IF_ERROR`

Gabriel39 commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,358 @@
+    // TODO: here calling add operator maybe only hava done one row, this performance may be poorly
+    // so it's possible to maintain a hashtable in FE, the key is place address, value is the object
+    // then we can calling add_bacth function and calculate the whole batch at once,
+    // and avoid calling jni multiple times.
```
Padding a space at the beginning for each commented line below `TODO`

Gabriel39 commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,358 @@
+    void add_batch_single_place(size_t batch_size, AggregateDataPtr place, const IColumn** columns,
+                                Arena* arena) const override {
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(_fn));
```
use `RETURN_IF_ERROR`

Gabriel39 replied:
Is it necessary to `init_udaf` for each time to call `add_batch_single_place`? Seems you have called it in `create`.

Gabriel39 commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,358 @@
+    std::unique_ptr<int64_t> output_offsets_ptr;
+    std::unique_ptr<int64_t> output_intermediate_state_ptr;
+
+    bool first_init;
```
I think this variable name is a bit curious. `init` also means `first`. Maybe rename to `inited` or `first_called` ?

Gabriel39 commented on be/src/vec/aggregate_functions/aggregate_function_java_udaf.h:
```diff
@@ -0,0 +1,358 @@
+        jbyteArray arr = (jbyteArray)(env->CallNonvirtualObjectMethod(executor_obj, executor_cl,
+                                                                      executor_serialize_id));
+        int len = env->GetArrayLength(arr);
+        serialize_data.resize(len);
```
call `serialize_data` before using it

HappenLee reviewed (approved):
LGTM
